sound/soc/sof/ipc4-pcm.c

```diff
@@ -91,17 +94,41 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	/*
 	 * IPC4 requires pipelines to be triggered in order starting at the sink and
 	 * walking all the way to the source. So traverse the pipeline_list in the reverse order.
-	 * Skip the pipelines that have their skip_during_fe_trigger flag set or if they're already
-	 * in the requested state. If there is a fork in the pipeline, the order of triggering
-	 * between the left/right paths will be indeterministic. But the sink->source trigger order
-	 * sink->source would still be guaranteed for each fork independently.
+	 * Skip the pipelines that have their skip_during_fe_trigger flag set. If there is a fork
+	 * in the pipeline, the order of triggering between the left/right paths will be
+	 * indeterministic. But the sink->source trigger order sink->source would still be
+	 * guaranteed for each fork independently.
 	 */
 	for (i = pipeline_list->count - 1; i >= 0; i--) {
 		spipe = pipeline_list->pipelines[i];
 		pipe_widget = spipe->pipe_widget;
 		pipeline = pipe_widget->private;
-		if (pipeline->state != state && !pipeline->skip_during_fe_trigger)
-			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
+
+		if (pipeline->skip_during_fe_trigger)
+			continue;
+
+		switch (state) {
+		case SOF_IPC4_PIPE_RUNNING:
+			/* Do not repeat RUNNING if the pipeline is already running */
+			if (spipe->started_count == spipe->paused_count)
+				data->pipeline_ids[data->count++] = pipe_widget->instance_id;
+			break;
+		case SOF_IPC4_PIPE_RESET:
+			/* RESET if the pipeline is in not in use, neither running nor paused */
```

Comment:
This is the state machine according to document.  if pipeline is paused it can be reset. The reset pipeline can also be paused
````c
 *                      INIT
 *                       |    \
 *                       |   __\|
 *                       |
 *                       |     RESET
 *                       |     _   _
 *                       |     /| |\
 *                       |    /    /\
 *                      \|/ |/_   /  \
 *        RUNNING <--> PAUSE _   /    \
 *            /  \      /|\ |\  /      \
 *           /    \      |    \/        \
 *          /      \     |    /\         \
 *         /        \    |   /  \         \
 *       |/_        _\|  |  /    \        _\|
 *     ERROR Stop       EOS       \______\ SAVE
 *                                      /
 ````

Reply:
@RanderWang Thanks but I don't think the driver needs to implement the state machine at all. The driver only cares about whether a pipeline should be paused or not or started or not depending on the  number of users and their states.

Reply:
yes, I had the same idea later.